### DISCUSSION TITLE
Attachment Searching, Colours and Patterns

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3228] [Development]
NeoMutt's colours work in two ways:

- Objects: Define some "things" that are coloured,
  e.g. `color status red blue`

- Regex: Colour any matching lines,
  e.g. `color index '~s neomutt' green default`

We need a list of every "object" and every "attribute" of an attachment,
e.g.
- `color attach_group green default`
  Highlight groups
- `color attach '~S >10K' white red`
  Highlight large attachments
 
It would also be good to be able to search using patterns. At present attachment search just matches whatever is actually visible as produced by `attach_format`.

---

Here's a sample email to consider:

```
- I   1 /tmp/xyz                                    [text/plain, 7bit, us-ascii, 1.0K]
  A   2 alternates.c                                [text/plain, 7bit, us-ascii, 4.3K]
  A   3 <no description>                             [multipart/alternative, 7bit, 0K]
  A   4 ├─>command_parse.c                           [text/plain, 7bit, us-ascii, 43K]
  A   5 ├─>commands.c                                [text/plain, 7bit, us-ascii, 34K]
  A   6 ├─>complete.c                               [text/plain, 7bit, us-ascii, 6.5K]
  A   7 ├─>conststrings.c                           [text/plain, 7bit, us-ascii, 5.9K]
  A   8 └─>context.c                                 [text/plain, 7bit, us-ascii, 10K]
  A   9 copy.c                                       [text/plain, 7bit, us-ascii, 32K]
```

**Comment by dcpurton** [comment]
## Attachment colour objects

### Likely useful

| Object | Pattern | Highlights |
|:---|:---:|:---|
| `attach` | yes | Entire attachment line |
| `attach_collapsed` | no | Number of attachments in collapsed group, `%N` |
| `attach_count` | no | Qualifies for counting, `%Q` |
| `attach_deleted` | no | Deleted flag, `%D` |
| `attach_name` | yes | Description or Filename, `%d` `%F` `%f` |
| `attach_multipart` | no | Multipart type flag, `%g` |
| `attach_number` | no | Number in attachment list, `%n` |
| `attach_unlink` | no | Unlink flag, `%u` |

### Possibly useful

| Object | Pattern | Highlights |
|:---|:---:|:---|
| `attach_charset` | no | Charset, `%C` |
| `attach_encoding` | no | Content transfer encoding, `%e` |
| `attach_id` | no | Content ID, `%i` |
| `attach_language` | no | Content language, `%l` |
| `attach_size` | no | Size, `%s` |
| `attach_subtype` | no | MIME subtype, `%M` |
| `attach_type` | no | MIME type, `%m` |

## New formatting sequences for attachment list

| Sequence | Description |
|:---:|:---|
| `%i` | Content ID |
| `%g` | Multipart group type flag (A, L, R, E, S, M) |
| `%l` | Content language |
| `%N` | Number of hidden parts in a collapsed attachment group |

## Attachment patterns (for colours and searching)

| Pattern modifier | Notes | Description |
|:---|:---|:---|
| `~A` | | all attachments |
| `~b EXPR` | |	attachments which contain `EXPR` in their contents |
| `~d EXPR` | | attachments with description matching `EXPR` |
| `~D` | | attachments marked to be deleted |
| `~f EXPR` | | attachments with filename matching `EXPR` |
| `~i EXPR` | |	attachments with a `Content-ID` matching `EXPR` |
| `~l EXPR` | | attachments with a `Content-Language` matching `EXPR` |
| `~k` | | attachments which contain PGP key material |
| `~n EXPR` | | attachments with filename or description matching `EXPR` |
| `~N [MIN]-[MAX]` | <sup>1</sup> | collapsed groups with between `MIN` and `MAX` parts |
| `~T` | | tagged attachments |
| `~u` | | attachments marked to be unlinked |
| `~X [MIN]-[MAX]` | <sup>1</sup> | attachments with `MIN` and `MAX` parts |
| `~z [MIN]-[MAX]` | <sup>1, 2</sup> | attachments with a size in the range `MIN` to `MAX` |
| `~(PATTERN)` | | attachments in groups containing attachments matching `PATTERN` |
| `~<(PATTERN)` | |	attachments whose immediate parent matches `PATTERN` |
| `~>(PATTERN)` | |	attachments having an immediate child matching `PATTERN` |

<sup>1</sup> The forms `<[MAX]`, `>[MIN]`, `[MIN]-` and `-[MAX]` are allowed, too.
<sup>2</sup> The suffixes `K` and `M` are allowed to specify kilobyte and megabyte respectively.

**WIP**

**Reply by dcpurton** [reply]
Note that `%N` will require `count_attachments()` to be moved out of `compose/functions.c` and made a public function.

**Reply by flatcap** [reply]
> `count_attachments()`

Yeah.  Even if it's only used in one place, it ought to live in libattach.

**Reply by flatcap** [reply]
I like the idea of default config:
```sh
# Warn about large attachments
color attach white red "~z >50k"
```

**Reply by dcpurton** [reply]
> > `count_attachments()`
> 
> Yeah. Even if it's only used in one place, it ought to live in libattach.

I wonder also about `find_body_parent()` and `find_body_previous()`.

All of these only depend on `struct Body`.

Which file and header should they go in?

And presumably they should be renamed with a prefix?

**Reply by flatcap** [reply]
When I created libattach, I just dumped `*attach*.[ch]` in there.
They desperately need a sort out, but I haven't investigated _how_.
Suggestions welcome!

I don't mind if we add another file, for now.
Maybe `lib.c` -- functions that only depend on `Body`

> `find_body_parent()` and `find_body_previous()`.
> renamed with a prefix?

I imagine that `Body` was named when it represented the single body of an email -- pre-mime.
'Attachment` is a bit specific.  Not all mime parts _are_ attachments.

As they're in libattach, let's go for `attach_*()` for now.
I'm happy to rename everything later, when we come up with something better.

**Reply by dcpurton** [reply]
OK. I'll see if I can put together a PR which moves these functions into `attach/lib.c`

**Reply by dcpurton** [reply]
> 'Attachment` is a bit specific.  Not all mime parts _are_ attachments.

At one stage I think I had `count_attachments()` called `body_count()`... But changed it on the grounds of seemed a bit macabre 😬

**Reply by dcpurton** [reply]
For reference, see commits 27f11326465f42270428dfd4bde211f114422be3 and 5ecb56e228bdb7404ea4df310572fb925fdcb834.
